File: Java/290. Word Pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        HashMap mapping = new HashMap();
        
        String[] words = s.split(" ");
        if(words.length != pattern.length()){
            return false;
        }
        
        for(Integer i=0; i<pattern.length(); i++){
            char c = pattern.charAt(i);
            String word = words[i];
            
            if(!mapping.containsKey(c)){
                mapping.put(c, i);
            }
            
            if(!mapping.containsKey(word)){
                mapping.put(word, i);
            }
            
            if(mapping.get(c) != mapping.get(word)){
                return false;
            }
        
        }
        
        return true;
    }
}

File: Python3/290. Word Pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        ptns = collections.defaultdict(list)
        
        for i, c in enumerate(pattern):
            ptns[c].append(i)
            
        s = s.split()
        
        if len(s) != len(pattern):
            return False
        
        # two chars cannot be mapped to the same word
        seen = set()
        
        for idx in ptns.values():
            p = s[idx[0]]
            if p in seen:
                return False
            
            seen.add(p)
            
            for i in range(1, len(idx)):
                if s[idx[i]] != p:
                    return False
                
        return True